### DISCUSSION TITLE
YARP: how to supply default credentials in a proxied request?

### DISCUSSION THREAD

**Discussion by MattiasMartens** [dotnet/yarp#1886] [Q&A] [answered]
Hi, I first [posted this on Stack Overflow](https://stackoverflow.com/questions/73959965/yarp-how-to-supply-default-credentials-in-a-proxied-request) but I thought it might be good to take it straight to the source as well.

Quoting from SO:

Per https://github.com/microsoft/reverse-proxy/issues/166#issuecomment-1267355264 it is possible to authenticate using the default network credentials cache when a server forwards a request through YARP.

(That is, it is possible to authenticate using credentials stored on the server; a user logging in through a proxy is confirmed not to be supported.)

But how to actually configure this?

The docs suggest that I can add a "WebProxy" block to my network configuration file, to which I can add a "UseDefaultCredentials" parameter:

```json
      "MyCluster": {
        "Destinations": {
          "mydest": {
            "Address": "https://example.com",
            "UseDefaultCredentials": "true"
          }
        },
        "HttpClient": {
          "WebProxy": {
            "Address": "??? what goes here?",
            "UseDefaultCredentials": "true"
          }
        }
      },
```

But the question is in the code: what exactly is the address supposed to be for the outgoing web proxy? Is this separate from the URLs in the destinations section? What if there is more than one destination?

I have tried:

- Specifying HttpClient.WebProxy.UseDefaultCredentials and leaving HttpClient.WebProxy.Address blank (not respected, still get a request for authentication)
- Specifying both HttpClient.WebProxy.UseDefaultCredentials and HttpClient.WebProxy.Address, removing the "Destinations" section entirely (503 Service not configured)

**Comment by samsp-msft** [comment, marked as answer]
The HttpClient.WebProxy is not for this scenario, it is for when using a forward proxy that you need to authenticate with, and will not affect the authentication between YARP and the destination server.

You probably need to be looking at [Configuring the client](https://microsoft.github.io/reverse-proxy/articles/http-client-config.html#configuring-the-http-client) and then setting a credential something like [this](https://weblog.west-wind.com/posts/2021/Nov/27/NTLM-Windows-Authentication-Authentication-with-HttpClient).

**Reply by MattiasMartens** [reply]
Ok, thanks, so, something like this?

```csharp
            var proxyBuilder = services
                .AddReverseProxy()
                .ConfigureHttpClient((context, handler) =>
                {
                    handler.Credentials = CredentialCache.DefaultNetworkCredentials;
                });
```

**Reply by samsp-msft** [reply]
Probably. That part of HttpClient is not my area of expertise. The thing to consider is what user account is YARP running under, as this would use that for the connection to the back end.

**Reply by mattiasgth** [reply]
@MattiasMartens Did you have any success with this? I'm in a similar situation, where the service running yarp should use Windows Authentication at the destination server, using the application pool identity (not the end-user's identity). I'm not sure if this is even possible using yarp, but it would simplify things greatly for us.

**Reply by Tratcher** [reply]
It should be possible using the approach Mattias showed above.

**Reply by mattiasgth** [reply]
@Tratcher: thanks for the encouragement :-)  I've tried all different kinds of tricks to get this working, but no luck so far. Using our own homegrown proxy (where a .NET HttpClient is sending the request), everything works as expected, but there's obviously something wrong with how we're using yarp. I'll start a new thread Monday on this. (It would be great to have an article on proxy => destination auth in the docs.)

**Reply by MattiasMartens** [reply]
@mattiasgth Hi, fellow Mattias.

I posted this on another page -- what I had above was a starting point but it would fail on execution because NTLM is a multi-step flow. The thing that ultimately worked for us was more like this:

```
// Add reverse proxy capability to the server
var proxyBuilder = services
    .AddReverseProxy()
    .ConfigureHttpClient((context, handler) =>
    {
        // Set credentials here
        handler.Credentials = CredentialCache.DefaultNetworkCredentials;
    })
    // Load configuration here
    .LoadFromConfig(Configuration.GetSection("..."))
    .AddTransforms(context => {
        // Load request content fully into buffer instead of providing it as a stream.
        // This is needed to prevent the multi-message NTLM auth flow from
        // prompting the request stream to be read more than once, which is an error.
        context.AddRequestTransform(async transformContext => {
            await transformContext.ProxyRequest.Content.LoadIntoBufferAsync();
        });
    });
```

**Reply by mattiasgth** [reply]
OK, so I never got this to work. I've tried adding the transform above, but no luck. Our backend server returns `The HTTP/2 server reset the stream. HTTP/2 error code 'HTTP_1_1_REQUIRED' (0xd).` when using this kind of authentication. Anonymous authentication works fine.

On what version of IIS did @MattiasMartens get this to work? Any ideas on how to debug this?

**Reply by Tratcher** [reply]
Oh, Windows auth doesn't work over HTTP/2 (YARPs default). You need to tell the proxy to use HTTP/1.1.

```
    "Clusters": {
      "allclusterprops": {
        "HttpRequest" : { // Options for sending request to destination
          "ActivityTimeout" : "00:02:00",
          "Version" : "1.1",
          "VersionPolicy" : "RequestVersionOrLower",
        },
....
      }
```

**Reply by mattiasgth** [reply]
@Tratcher Many thanks for this! Sorry for very late answer, but switching to HTTP/1.1 _almost_ solved it. Everything works as expected in all environments, except our production servers. For some reason, the requests fail when using HTTPS. What ultimately worked for us was to switch to plain HTTP for proxy <=> back-end communication.
